src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs

```diff
@@ -372,115 +374,142 @@ private bool RunTestInternalWithExecutors(IEnumerable<Tuple<Uri, string>> execut
             // Call the executor for each group of tests.
             var exceptionsHitDuringRunTests = false;
 
-            // Collecting Total Number of Adapters Discovered in Machine
+            // Collecting Total Number of Adapters Discovered in Machine.
             this.requestData.MetricsCollection.Add(TelemetryDataConstants.NumberOfAdapterDiscoveredDuringExecution, executorUriExtensionMap.Count());
 
+            bool attachedToTestHost = false;
+            var executorCache = new List<LazyExtension<ITestExecutor, ITestExecutorCapabilities>>();
             foreach (var executorUriExtensionTuple in executorUriExtensionMap)
             {
-                // Get the executor
+                // Get the executor.
                 var extensionManager = this.GetExecutorExtensionManager(executorUriExtensionTuple.Item2);
 
                 // Look up the executor.
                 var executor = extensionManager.TryGetTestExtension(executorUriExtensionTuple.Item1);
-                if (executor != null)
+
+                // Cache the executor.
+                executorCache.Add(executor);
+
+                // Check if we actually have to attach to the default test host.
+                if (!this.runContext.IsBeingDebugged || attachedToTestHost)
                 {
-                    try
+                    // We already know we should attach to the default test host, simply continue.
+                    continue;
+                }
+
+                if (!(executor.Value is ITestExecutor2) || this.ShouldAttachToTestHost(executor.Value))
+                {
+                    EqtTrace.Verbose("Attaching to default test host.");
+
+                    attachedToTestHost = true;
+                    if (!this.frameworkHandle.AttachDebuggerToProcess(Process.GetCurrentProcess().Id))
                     {
-                        if (EqtTrace.IsVerboseEnabled)
-                        {
-                            EqtTrace.Verbose(
-                                "BaseRunTests.RunTestInternalWithExecutors: Running tests for {0}",
-                                executor.Metadata.ExtensionUri);
-                        }
+                        throw new TestPlatformException("Cannot attach the debugger to the default test host.");
+                    }
+                }
+            }
 
-                        // set the active executor
-                        this.activeExecutor = executor.Value;
+            int index = 0;
+            foreach (var executorUriExtensionTuple in executorUriExtensionMap)
+            {
+                // Get the executor from cache.
+                var executor = executorCache[index++];
```

Comment:
Took care of the order dependency issue by having the `executorCache` be a `Dictionary<K, V>`.

However we're still iterating the `executorUriExtensionMap` (because we need that info and it's not stored in the `executorCache`) and get the corresponding executor by searching the dictionary.